External pipe patch is added

Abhra00 opened the pull request.

Abhra00 commented:
Externalpipe patch merged to main brach of my fork

DreamMaoMao commented:
What is the function of this patch? Is it to open the connection and copy the screen data through regex?

Abhra00 commented:
> What is the function of this patch? Is it to open the connection and copy the screen data through regex?

yes,
I have added two functionalities where if you press alt-l then a dmenu menu will pop-up containing all urls within current buffer and if you select one then the url will be opened in your default browser 

and 
 if you press alt-o then  a dmenu menu will pop-up  containing all commands with current buffer and you can copy the output of your desired command

Abhra00 commented:
> What is the function of this patch? Is it to open the connection and copy the screen data through regex?
> 
> if so , it already support in #20.

sir if you have could checkout this functionality by compiling it 

it is in my github
https://github.com/Abhra00/st-sx.git

veltza commented on config.def.h:
```diff
@@ -419,6 +419,12 @@ static MouseShortcut mshortcuts[] = {
+static char *openurlcmd[]    = { "/bin/sh", "-c", "st-urlhandler -o", "externalpipe", NULL };
+static char *copyurlcmd[]    = { "/bin/sh", "-c", "st-urlhandler -c", "externalpipe", NULL };
+static char *copyoutput[]    = { "/bin/sh", "-c", "st-copyout", "externalpipe", NULL };
+static char *setbgcolorcmd[] = { "/bin/sh", "-c", "printf '\033]11;#008000\007'", "externalpipein", NULL };
```
Comment out this line as it is not bound to any shortcut and is just an example of a possible use case.

veltza commented on patch/externalpipe.c:
```diff
@@ -0,0 +1,65 @@
+	/* ignore sigpipe for now, in case child exists early */
+	oldsigpipe = signal(SIGPIPE, SIG_IGN);
+	newline = 0;
+	for (n = 0; n < term.row; n++) {
+		bp = term.line[n];
+		lastpos = MIN(tlinelen(TLINE(n)) + 1, term.col) - 1;
+		if (lastpos < 0)
+			break;
+		end = &bp[lastpos + 1];
+		for (; bp < end; ++bp)
+			if (xwrite(to[1], buf, utf8encode(bp->u, buf)) < 0)
+				break;
+		if ((newline = term.line[n][lastpos].mode & ATTR_WRAP))
+			continue;
+		if (xwrite(to[1], "\n", 1) < 0)
+			break;
+		newline = 0;
+	}
```
This should pipe both the screen and scrollback buffers.

veltza commented on st-copyout:
```diff
@@ -0,0 +1,13 @@
+sed -n "w $tmpfile"
+sed -i 's/\x0//g' "$tmpfile"
+ps1="$(grep "\S" "$tmpfile" | tail -n 1 | sed 's/^\s*//' | cut -d' ' -f1)"
+chosen="$(grep -F "$ps1" "$tmpfile" | sed '$ d' | tac | dmenu -p "Copy which command's output?" -i -l 10 -X 645 -Y 370 -W 580 | sed 's/[^^]/[&]/g; s/\^/\\^/g')"
```
This should work with an unpatched dmenu.

veltza commented on st-urlhandler:
```diff
@@ -0,0 +1,19 @@
+
+while getopts "hoc" o; do case "${o}" in
+	h) printf "Optional arguments for custom use:\\n  -c: copy\\n  -o: xdg-open\\n  -h: Show this message\\n" && exit 1 ;;
+	o) chosen="$(echo "$urls" | dmenu -i -p 'Follow which url?' -l 10 -X 645 -Y 370 -W 580)"
+	setsid xdg-open "$chosen" >/dev/null 2>&1 & ;;
+	c) echo "$urls" | dmenu -i -p 'Copy which url?' -l 10 -X 645 -Y 370 -W 580 | tr -d '\n' | xclip -selection clipboard ;;
```
These should work with an unpatched dmenu.

veltza reviewed (changes requested):
The externapipe patch does not provide any added value as it is, because it only operates with the screen buffer and the same can be done using the keyboard selection mode without external scripts. For example:

1. You can use the Ctrl+Shift+M/N shortcuts to highlight URLs and open or copy them to the clipboard.

2. If your shell prompt emits the OSC-133 sequence, you can activate keyboard selection mode by pressing Ctrl+Shift+Space. Then use the Z key to jump to the command you need, enter line selection mode with V, jump to the next command with X, and finally use Enter to copy the command output to the clipboard. It works, but I admit it's more clumsy to use for this purpose than externalpipe.

However, if you could modify externalpipe so that it would pipe both the screen and scrollback buffers to external scripts, then it would be much more useful. Because then you could open or copy any URL from the scrollback, which the keyboard selection mode cannot do because it only operates with visible URLs.

The last thing I would want from this patch is for the scripts to work with an unpatched dmenu, because otherwise users will file unnecessary bug reports.

So the bottom line is that I see a lot of uses for the patch even though it has some overlapping uses with the keyboard selection mode. But the above changes are needed.

Abhra00 commented:
Okay sir
You are wanting to pipe out info from whole scrollback buffer and the script should work with unpatched dmenu

veltza commented:
> Okay sir You are wanting to pipe out info from whole scrollback buffer and the script should work with unpatched dmenu

Exactly. If you need any help, just ask.

Abhra00 commented:
> > Okay sir You are wanting to pipe out info from whole scrollback buffer and the script should work with unpatched dmenu
> 
> Exactly. If you need any help, just ask.

Sir I am just a college student

I will try my best 

Thank you for your time sir

veltza commented:
If you want, I can fix the piping thing if you fix the rest? There is no hurry.

Abhra00 commented:
> If you want, I can fix the piping thing if you fix the rest? There is no hurry.

Okay sir

Abhra00 commented:
> If you want, I can fix the piping thing if you fix the rest? There is no hurry.

sir I have done the thing you have wanted 
could you please check it out

DreamMaoMao commented:
https://github.com/user-attachments/assets/0d7561d4-59b4-48ae-8f3c-6f0784b54351

It seems that it cannot copy the command correctly.

If wants to copy historical commands, I think atuin is a better solution.
![image](https://github.com/user-attachments/assets/a034f16a-8c6a-45d9-b4af-aed9d60e9698)

Abhra00 commented:
> 2025-04-19.14-06-33.mp4
> 
> It seems that it cannot copy the command correctly.
> 
> If wants to copy historical commands, I think atuin is a better solution. ![image](https://private-user-images.githubusercontent.com/30348075/435364437-a034f16a-8c6a-45d9-b4af-aed9d60e9698.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUwNDMyNjcsIm5iZiI6MTc0NTA0Mjk2NywicGF0aCI6Ii8zMDM0ODA3NS80MzUzNjQ0MzctYTAzNGYxNmEtOGM2YS00NWQ5LWI0YWYtYWVkOWQ2MGU5Njk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE5VDA2MDkyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYyNDZhMTc0YmY0YTExNTJmYWJiYWQzMWVmYzdjMmYyYzA2NmY2NjZhODE3OWU0ZmMyMjUwYjRmNDI4Nzg3ZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ev8EPTw-SkSsvKee6iokEjvquPh5vIgHMHHt5gVZMKc)

what is the problem you are having with copying commands

DreamMaoMao commented:
you can check the video ,  i copy some string that is not a command ,  and  copy with a wrong format.

Abhra00 commented:
> 2025-04-19.14-02-39.mp4
> 
> It seems that it cannot copy the command correctly.
> 
> If wants to copy historical commands, I think atuin is a better solution. ![image](https://private-user-images.githubusercontent.com/30348075/435364437-a034f16a-8c6a-45d9-b4af-aed9d60e9698.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUwNDMzMjMsIm5iZiI6MTc0NTA0MzAyMywicGF0aCI6Ii8zMDM0ODA3NS80MzUzNjQ0MzctYTAzNGYxNmEtOGM2YS00NWQ5LWI0YWYtYWVkOWQ2MGU5Njk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE5VDA2MTAyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ4YzIxNDk1NTljZTMwNWQ4YjQyMjk3NGQyYjg3YTJhYzQyOTU4MDljNmQ5NjEyYmExNGRmYjc0NDFiNjFlYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6WXrw7lc1ohmMJ34T6VgBdOk-S4n5zaw1EeC5Ks2nhI)

the problem you are showing is not happening on my side

DreamMaoMao commented:
well. It seems that what I copied was the output not cmd, but what I pasted was garbled.

Abhra00 commented:
> well. It seems that what I copied was the output not cmd, but what I pasted was garbled.

yes it is intended for copying the output of the command
but don't know why it is pasting garbled

Abhra00 commented:
> well. It seems that what I copied was the output not cmd, but what I pasted was garbled.

could you please check how the url handling is performing

DreamMaoMao commented:
The command to copy cannot be detected. The copy url and jump url are normal. The garbled characters might be due to the incorrect handling of some formatted strings when copying the buffer, such as the output of lsd -l

Abhra00 commented:
> The command to copy cannot be detected. The copy url and jump url are normal. The garbled characters might be due to the incorrect handling of some formatted strings when copying the buffer, such as the output of lsd -l

Sir can you fix the garbled character thing
And I think for most of the cases the string formatting will not be a big issue

DreamMaoMao commented:
I don't know how to fix it either, but kitty provides an option for pipe output to remove the format characters of the string before outputting it, i think should be do same thing here.

Abhra00 commented:
> I don't know how to fix it either, but kitty provides an option for pipe output to remove the format characters of the string before outputting it, i think should do same thing here.

Copying output mainly done by shell script st-copyoutpt can you check that how things should be done

veltza commented:
> sir I have done the thing you have wanted
> could you please check it out

@Abhra00 The scripts are now fine, but the other solution is not correct for the these reasons:

1. You are copying the entire history buffer even though only part of it has been used. This is why it can copy garbage too.

2. When the alternate screen is enabled, history should NOT be copied at all.

I think it would be better if I handled this task.

> If wants to copy historical commands, I think atuin is a better solution.

@DreamMaoMao The st-copyout script copies the output of previous commands, so it doesn't try to be atuin. I tested it yesterday and it seemed to work, but now it's broken because the latest commit doesn't work correctly. The script was written by Luke Smith and his friends, and it is a fairly old script, so it is possible that it does not work with every shell prompt and for that reason, it may not find the commands.

Abhra00 commented:
> > sir I have done the thing you have wanted
> > could you please check it out
> 
> @Abhra00 The scripts are now fine, but the other solution is not correct for the these reasons:
> 
> 1. You are copying the entire history buffer even though only part of it has been used. This is why it can copy garbage too.
> 2. When the alternate screen is enabled, history should NOT be copied at all.
> 
> I think it would be better if I handled this task.
> 
> > If wants to copy historical commands, I think atuin is a better solution.
> 
> @DreamMaoMao The st-copyout script copies the output of previous commands to the clipboard, so it doesn't try to be atuin. I tested it yesterday and it seemed to work, but now it's broken because the latest commit doesn't work correctly. The script was written by Luke Smith and his friends, and it is a fairly old script, so it is possible that it does not work with every shell prompt and for that reason, it may not find the commands.

Okay sir u can do the pipe out
Tell me sir if there is any help needed

veltza reviewed (approved):
I fixed the pipe issues and simplified the st-urlhandler script a bit, as the previous version that removed mutt sidebars broke url detection if mutt was not used.

I'll leave this PR open for a while for testing and feedback.

Abhra00 commented:
> I fixed the pipe issues and simplified the st-urlhandler script a bit, as the previous version that removed mutt sidebars broke url detection if mutt was not used.
> 
> I'll leave this PR open for a while for testing and feedback.

Okay sir 
I will try it out
And I am basically a noob at this type of programming 
But thank you sir for pointing out mistakes

Abhra00 commented:
> > I fixed the pipe issues and simplified the st-urlhandler script a bit, as the previous version that removed mutt sidebars broke url detection if mutt was not used.
> > I'll leave this PR open for a while for testing and feedback.
> 
> Okay sir I will try it out And I am basically a noob at this type of programming But thank you sir for pointing out mistakes

sir i have checked it out 
it is working perfectly fine

veltza commented:
I've been thinking about the `st-urlhandler` script and have come to the conclusion that I wouldn't want to add it to this project for the following reasons:

* We already support two options (mouse and keyboard) to open/copy links, so do we really need a third one?
* It doesn't work with hyperlinks because externalpipe doesn't output them. And I'm not sure how that should be implemented.
* I don't need it, so I wouldn't want to maintain it either.
* If someone needs it, they can use and maintain their own script.

Are there any strong reasons why it should be added or why that workflow would be better than the current ones?

Abhra00 commented:
> I've been thinking about the `st-urlhandler` script and have come to the conclusion that I wouldn't want to add it to this project for the following reasons:
> 
> * We already support two options (mouse and keyboard) to open/copy links, so do we really need a third one?
> * It doesn't work with hyperlinks because externalpipe doesn't output them. And I'm not sure how that should be implemented.
> * I don't need it, so I wouldn't want to maintain it either.
> * If someone needs it, they can use and maintain their own script.
> 
> Are there any strong reasons why it should be added or why that workflow would be better than the current ones?

Keeping urlhandler aside 
The copyoutput script is really helpful
Because you only need to press one key and that's it
Now I am not sure about the url-handler script

veltza commented:
Merged without the st-urlhandler script.

Thanks for your contribution!

Abhra00 commented:
> Merged without the st-urlhandler script.
> 
> Thanks for your contribution!

Thank you sir for helping me
